lib/irq.c

```diff
@@ -154,6 +154,34 @@ irqs_reset(vfu_ctx_t *vfu_ctx)
     }
 }
 
+static void
+irqs_set_state(vfu_ctx_t *vfu_ctx, struct vfio_irq_set *irq_set)
+{
+    vfu_dev_irq_state_cb_t *cb = NULL;
+    uint32_t irq_action, i;
+    bool mask = false;
+
+    assert(irq_set->index < VFU_DEV_NUM_IRQS);
+    cb = vfu_ctx->irq_state_cbs[irq_set->index];
+    if (!cb) {
+        return;
+    }
+
+    assert((irq_set->start + irq_set->count) <=
+               vfu_ctx->irq_count[irq_set->index]);
+
+    irq_action = irq_set->flags & VFIO_IRQ_SET_ACTION_TYPE_MASK;
+
+    assert((irq_action & VFIO_IRQ_SET_ACTION_MASK) ||
+           (irq_action & VFIO_IRQ_SET_ACTION_UNMASK));
+
+    mask = (irq_action & VFIO_IRQ_SET_ACTION_MASK) ? true : false;
+
+    for (i = irq_set->start; i < irq_set->count; i++) {
+        cb(vfu_ctx, i, mask);
```

Comment:
I think both of them should be OK.

I chose calling multiple times because that's how the guest OS does it. It writes to the register of vector iteratively.

It doesn't add any overhead because the library is baked into the device/server?

Reply:
if you provide the whole range, a callback implementation can easily do it one-by-one.
If you do it one-by-one, a callback implementation *couldn't* handle the whole range.

So it's more flexible to provide the information.

Reply:
OK, we could provide the whole range. Will post a new rev shortly.